File: opaque/src/main/java/com/codeheadsystems/opaque/internal/OpaqueCredentials.java
package com.codeheadsystems.opaque.internal;

import com.codeheadsystems.ellipticcurve.curve.OctetStringUtils;
import com.codeheadsystems.opaque.config.OpaqueCipherSuite;
import com.codeheadsystems.opaque.config.OpaqueConfig;
import com.codeheadsystems.opaque.internal.OpaqueEnvelope.RecoverResult;
import com.codeheadsystems.opaque.internal.OpaqueEnvelope.StoreResult;
import com.codeheadsystems.opaque.model.ClientRegistrationState;
import com.codeheadsystems.opaque.model.CredentialRequest;
import com.codeheadsystems.opaque.model.CredentialResponse;
import com.codeheadsystems.opaque.model.Envelope;
import com.codeheadsystems.opaque.model.RegistrationRecord;
import com.codeheadsystems.opaque.model.RegistrationRequest;
import com.codeheadsystems.opaque.model.RegistrationResponse;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;

/**
 * Handles the credential request/response lifecycle for OPAQUE registration and authentication.
 */
public class OpaqueCredentials {

  private OpaqueCredentials() {
  }

  /**
   * Client creates a registration request by blinding the password.
   */
  public static ClientRegistrationState createRegistrationRequest(byte[] password, OpaqueConfig config) {
    BigInteger blind = config.cipherSuite().oprfSuite().randomScalar();
    return createRegistrationRequestWithBlind(password, blind, config);
  }

  /**
   * Client creates a registration request with a given blinding factor (for deterministic testing).
   */
  public static ClientRegistrationState createRegistrationRequestWithBlind(byte[] password,
                                                                           BigInteger blind,
                                                                           OpaqueConfig config) {
    byte[] blindedElement = OpaqueOprf.blind(config.cipherSuite(), password, blind);
    RegistrationRequest request = new RegistrationRequest(blindedElement);
    return new ClientRegistrationState(blind, password, request);
  }

  /**
   * Server creates a registration response: evaluates OPRF and includes server public key.
   */
  public static RegistrationResponse createRegistrationResponse(OpaqueConfig config,
                                                                RegistrationRequest request,
                                                                byte[] serverPublicKey,
                                                                byte[] credentialIdentifier,
                                                                byte[] oprfSeed) {
    BigInteger oprfKey = OpaqueOprf.deriveOprfKey(config.cipherSuite(), oprfSeed, credentialIdentifier);
    byte[] evaluatedElement = OpaqueOprf.blindEvaluate(config.cipherSuite(), oprfKey, request.blindedElement());
    return new RegistrationResponse(evaluatedElement, serverPublicKey);
  }

  /**
   * Client finalizes registration: derives randomized_pwd, stores envelope.
   */
  public static RegistrationRecord finalizeRegistration(ClientRegistrationState state,
                                                        RegistrationResponse response,
                                                        byte[] serverIdentity,
                                                        byte[] clientIdentity,
                                                        OpaqueConfig config) {
    return finalizeRegistrationWithNonce(state, response, serverIdentity, clientIdentity,
        config, config.randomConfig().randomBytes(OpaqueConfig.Nn));
  }

  /**
   * Client finalizes registration with a provided nonce (for deterministic testing).
   */
  public static RegistrationRecord finalizeRegistrationWithNonce(ClientRegistrationState state,
                                                                 RegistrationResponse response,
                                                                 byte[] serverIdentity,
                                                                 byte[] clientIdentity,
                                                                 OpaqueConfig config,
                                                                 byte[] envelopeNonce) {
    byte[] randomizedPwd = deriveRandomizedPwd(state.password(), state.blind(),
        response.evaluatedElement(), config);

    StoreResult stored = OpaqueEnvelope.store(config,
        randomizedPwd, response.serverPublicKey(), serverIdentity, clientIdentity, envelopeNonce);
    return new RegistrationRecord(stored.clientPublicKey(), stored.maskingKey(), stored.envelope());
  }

  /**
   * Server creates a credential response for authentication.
   */
  public static CredentialResponse createCredentialResponse(OpaqueConfig config,
                                                            CredentialRequest request,
                                                            byte[] serverPublicKey,
                                                            RegistrationRecord record,
                                                            byte[] credentialIdentifier,
                                                            byte[] oprfSeed) {
    return createCredentialResponseWithNonce(config, request, serverPublicKey, record,
        credentialIdentifier, oprfSeed, config.randomConfig().randomBytes(OpaqueConfig.Nn));
  }

  /**
   * Server creates a credential response with a provided masking nonce (for deterministic testing).
   */
  public static CredentialResponse createCredentialResponseWithNonce(OpaqueConfig config,
                                                                     CredentialRequest request,
                                                                     byte[] serverPublicKey,
                                                                     RegistrationRecord record,
                                                                     byte[] credentialIdentifier,
                                                                     byte[] oprfSeed,
                                                                     byte[] maskingNonce) {
    OpaqueCipherSuite suite = config.cipherSuite();
    BigInteger oprfKey = OpaqueOprf.deriveOprfKey(suite, oprfSeed, credentialIdentifier);
    byte[] evaluatedElement = OpaqueOprf.blindEvaluate(suite, oprfKey, request.blindedElement());

    // pad = HKDF-Expand(masking_key, masking_nonce || "CredentialResponsePad", Npk + Nn + Nm)
    byte[] padInfo = OctetStringUtils.concat(
        maskingNonce,
        "CredentialResponsePad".getBytes(StandardCharsets.US_ASCII)
    );
    byte[] pad = OpaqueCrypto.hkdfExpand(suite, record.maskingKey(), padInfo, config.maskedResponseSize());

    // plaintext = server_public_key || envelope_nonce || auth_tag
    byte[] plaintext = OctetStringUtils.concat(serverPublicKey, record.envelope().serialize());
    byte[] maskedResponse = OpaqueCrypto.xor(pad, plaintext);

    return new CredentialResponse(evaluatedElement, maskingNonce, maskedResponse);
  }

  /**
   * Client recovers credentials from the credential response during authentication.
   */
  public static RecoverResult recoverCredentials(byte[] password, BigInteger blind,
                                                 CredentialResponse response,
                                                 byte[] serverIdentity,
                                                 byte[] clientIdentity,
                                                 OpaqueConfig config) {
    OpaqueCipherSuite suite = config.cipherSuite();
    byte[] randomizedPwd = deriveRandomizedPwd(password, blind, response.evaluatedElement(), config);

    // Recover masking_key = Expand(randomized_pwd, "MaskingKey", Nh)
    byte[] maskingKey = OpaqueCrypto.hkdfExpand(suite, randomizedPwd,
        "MaskingKey".getBytes(StandardCharsets.US_ASCII), config.Nh());

    // Unmask: pad = HKDF-Expand(masking_key, masking_nonce || "CredentialResponsePad", ...)
    byte[] padInfo = OctetStringUtils.concat(
        response.maskingNonce(),
        "CredentialResponsePad".getBytes(StandardCharsets.US_ASCII)
    );
    byte[] pad = OpaqueCrypto.hkdfExpand(suite, maskingKey, padInfo, config.maskedResponseSize());
    byte[] plaintext = OpaqueCrypto.xor(pad, response.maskedResponse());

    // Extract server_public_key || envelope
    byte[] serverPublicKey = new byte[config.Npk()];
    System.arraycopy(plaintext, 0, serverPublicKey, 0, config.Npk());
    Envelope envelope = Envelope.deserialize(plaintext, config.Npk(), OpaqueConfig.Nn, config.Nm());

    return OpaqueEnvelope.recover(config, randomizedPwd, serverPublicKey, envelope, serverIdentity, clientIdentity);
  }

  /**
   * Derives randomized password from OPRF output.
   */
  public static byte[] deriveRandomizedPwd(byte[] password, BigInteger blind,
                                           byte[] evaluatedElement, OpaqueConfig config) {
    byte[] oprfOutput = OpaqueOprf.finalize(config.cipherSuite(), password, blind, evaluatedElement);
    byte[] stretchedOutput = config.ksf().stretch(oprfOutput, config);
    return OpaqueCrypto.hkdfExtract(config.cipherSuite(), new byte[0],
        OctetStringUtils.concat(oprfOutput, stretchedOutput));
  }
}

File: hofmann-rust/src/opaque/internal/opaque_credentials.rs
use crate::common::{concat, xor};
use crate::opaque::config::OpaqueConfig;
use crate::opaque::config::NN;
use crate::opaque::internal::opaque_envelope::{self, RecoverResult};
use crate::opaque::internal::opaque_oprf;
use crate::opaque::model::*;

/// Client creates a registration request by blinding the password.
pub fn create_registration_request(
    password: &[u8],
    config: &OpaqueConfig,
    rng: &mut dyn rand_core::CryptoRngCore,
) -> ClientRegistrationState {
    let blind = config.cipher_suite().oprf_suite().random_scalar(rng);
    create_registration_request_with_blind(password, &blind, config)
}

/// Client creates a registration request with a given blinding factor (for deterministic testing).
pub fn create_registration_request_with_blind(
    password: &[u8],
    blind: &[u8],
    config: &OpaqueConfig,
) -> ClientRegistrationState {
    let blinded_element = opaque_oprf::blind(config.cipher_suite(), password, blind);
    ClientRegistrationState {
        blind: blind.to_vec(),
        password: password.to_vec(),
        request: RegistrationRequest { blinded_element },
    }
}

/// Server creates a registration response: evaluates OPRF and includes server public key.
pub fn create_registration_response(
    config: &OpaqueConfig,
    request: &RegistrationRequest,
    server_public_key: &[u8],
    credential_identifier: &[u8],
    oprf_seed: &[u8],
) -> Result<RegistrationResponse, &'static str> {
    let oprf_key =
        opaque_oprf::derive_oprf_key(config.cipher_suite(), oprf_seed, credential_identifier);
    let evaluated_element =
        opaque_oprf::blind_evaluate(config.cipher_suite(), &oprf_key, &request.blinded_element)?;
    Ok(RegistrationResponse {
        evaluated_element,
        server_public_key: server_public_key.to_vec(),
    })
}

/// Client finalizes registration: derives randomized_pwd, stores envelope.
pub fn finalize_registration(
    state: &ClientRegistrationState,
    response: &RegistrationResponse,
    server_identity: Option<&[u8]>,
    client_identity: Option<&[u8]>,
    config: &OpaqueConfig,
    rng: &mut dyn rand_core::CryptoRngCore,
) -> Result<RegistrationRecord, &'static str> {
    let mut nonce = vec![0u8; NN];
    rng.fill_bytes(&mut nonce);
    finalize_registration_with_nonce(
        state,
        response,
        server_identity,
        client_identity,
        config,
        &nonce,
    )
}

/// Client finalizes registration with a provided nonce (for deterministic testing).
pub fn finalize_registration_with_nonce(
    state: &ClientRegistrationState,
    response: &RegistrationResponse,
    server_identity: Option<&[u8]>,
    client_identity: Option<&[u8]>,
    config: &OpaqueConfig,
    envelope_nonce: &[u8],
) -> Result<RegistrationRecord, &'static str> {
    let randomized_pwd = derive_randomized_pwd(
        &state.password,
        &state.blind,
        &response.evaluated_element,
        config,
    )?;

    let stored = opaque_envelope::store(
        config,
        &randomized_pwd,
        &response.server_public_key,
        server_identity,
        client_identity,
        envelope_nonce,
    );

    Ok(RegistrationRecord {
        client_public_key: stored.client_public_key,
        masking_key: stored.masking_key,
        envelope: stored.envelope,
    })
}

/// Server creates a credential response for authentication.
pub fn create_credential_response_with_nonce(
    config: &OpaqueConfig,
    request: &CredentialRequest,
    server_public_key: &[u8],
    record: &RegistrationRecord,
    credential_identifier: &[u8],
    oprf_seed: &[u8],
    masking_nonce: &[u8],
) -> Result<CredentialResponse, &'static str> {
    let oprf_key =
        opaque_oprf::derive_oprf_key(config.cipher_suite(), oprf_seed, credential_identifier);
    let evaluated_element =
        opaque_oprf::blind_evaluate(config.cipher_suite(), &oprf_key, &request.blinded_element)?;

    // pad = HKDF-Expand(masking_key, masking_nonce || "CredentialResponsePad", Npk + Nn + Nm)
    let pad_info = concat(&[masking_nonce, b"CredentialResponsePad"]);
    let pad = config.cipher_suite().hkdf_expand(
        &record.masking_key,
        &pad_info,
        config.masked_response_size(),
    );

    // plaintext = server_public_key || envelope_nonce || auth_tag
    let plaintext = concat(&[server_public_key, &record.envelope.serialize()]);
    let masked_response = xor(&pad, &plaintext);

    Ok(CredentialResponse {
        evaluated_element,
        masking_nonce: masking_nonce.to_vec(),
        masked_response,
    })
}

/// Client recovers credentials from the credential response during authentication.
pub fn recover_credentials(
    password: &[u8],
    blind: &[u8],
    response: &CredentialResponse,
    server_identity: Option<&[u8]>,
    client_identity: Option<&[u8]>,
    config: &OpaqueConfig,
) -> Result<RecoverResult, &'static str> {
    let randomized_pwd =
        derive_randomized_pwd(password, blind, &response.evaluated_element, config)?;

    // Recover masking_key = Expand(randomized_pwd, "MaskingKey", Nh)
    let masking_key =
        config
            .cipher_suite()
            .hkdf_expand(&randomized_pwd, b"MaskingKey", config.nh());

    // Unmask
    let pad_info = concat(&[&response.masking_nonce, b"CredentialResponsePad"]);
    let pad =
        config
            .cipher_suite()
            .hkdf_expand(&masking_key, &pad_info, config.masked_response_size());
    let plaintext = xor(&pad, &response.masked_response);

    // Extract server_public_key || envelope
    let server_public_key = plaintext[..config.npk()].to_vec();
    let envelope = Envelope::deserialize(&plaintext, config.npk(), NN, config.nm());

    opaque_envelope::recover(
        config,
        &randomized_pwd,
        &server_public_key,
        &envelope,
        server_identity,
        client_identity,
    )
}

/// Derives randomized password from OPRF output.
///
/// Returns `Err` if `evaluated_element` (from the server) is not a valid
/// non-identity group element.
pub fn derive_randomized_pwd(
    password: &[u8],
    blind: &[u8],
    evaluated_element: &[u8],
    config: &OpaqueConfig,
) -> Result<Vec<u8>, &'static str> {
    let oprf_output =
        opaque_oprf::finalize(config.cipher_suite(), password, blind, evaluated_element)?;
    let stretched_output = config.stretch_password(&oprf_output);
    Ok(config
        .cipher_suite()
        .hkdf_extract(&[], &concat(&[&oprf_output, &stretched_output])))
}
